[ja] Update content/ja/docs/concepts/workloads/controllers/deployment.md (#56213)
kubernetes/website: content/ja/docs/concepts/workloads/controllers/deployment.md

    進化を続けるDeploymentのエコシステムを活用してください。
description: >-
  Deploymentは、アプリケーションのワークロード(通常は状態を保持しないもの)を実行するための一連のPodを管理します。
content_type: concept
weight: 10
hide_summary: true # セクションのインデックスに個別に掲載される
---

<!-- overview -->

_Deployment_ は{{< glossary_tooltip text="Pod" term_id="pod" >}}と{{< glossary_tooltip term_id="replica-set" text="ReplicaSet" >}}の宣言的なアップデート機能を提供します。

Deploymentにおいて _理想的な状態_ を記述すると、Deployment{{< glossary_tooltip text="コントローラー" term_id="controller" >}}は現在の状態を制御しながら理想的な状態に変更していきます。
Deploymentを定義することによって、新しいReplicaSetを作成したり、既存のDeploymentを削除し、その全てのリソースを新しいDeploymentに引き継いだりできます。

{{< note >}}
Deploymentによって所有されているReplicaSetを管理しないでください。
ご自身のユースケースが以下の項目に含まれない場合、メインのKubernetesリポジトリにIssueを作成することを検討してください。
{{< /note >}}

<!-- body -->

## ユースケース {#use-case}

以下の項目はDeploymentの典型的なユースケースです:

* [ReplicaSetをロールアウトするためにDeploymentの作成を行う](#creating-a-deployment)。
  ReplicaSetはバックグラウンドでPodを作成します。
  Podの作成が完了したかどうかは、ロールアウトのステータスを確認してください。
* DeploymentのPodTemplateSpecを更新することにより[Podの新しい状態を宣言する](#updating-a-deployment)。
  新しいReplicaSetが作成され、Deploymentは古いReplicaSetをスケールダウンしながら新しいReplicaSetを段階的にスケールアップすることで、Podを制御しながら置き換えます。
  新しいReplicaSetはそれぞれDeploymentのリビジョンを更新します。

> * Deploymentの現在の状態が不安定な場合、[Deploymentのロールバック](#rolling-back-a-deployment)をする。

Reviewer: ```suggestion
* Deploymentの現在の状態が不安定な場合、[過去のDeploymentのリビジョンにロールバックする](#rolling-back-a-deployment)。
```
原文のリンクテキストの「Rollback to an earlier Deployment revision」に合わせています。

Reviewer: 一度コメントをしたあとに、他の箇所での表現に合わせて、「以前の」を「過去の」に変えました。
すみませんが、改めて変更をお願いします。

  ロールバックによる各更新作業は、Deploymentのリビジョンを更新します。
* [より多くの負荷をさばけるように、Deploymentをスケールアップする](#scaling-a-deployment)。

> * PodTemplateSpecに対する複数の修正を適用するために[Deploymentを停止(Pause)し](#pausing-and-resuming-a-deployment)、それを再開して新しいロールアウトを開始します。

Reviewer: ```suggestion
* PodTemplateSpecに対する複数の修正を適用するために[Deploymentのロールアウトを一時停止(Pause)し](#pausing-and-resuming-a-deployment)、それを再開して新しいロールアウトを開始する。
```
このセクションでの箇条書きの1文目は、「○○する」で統一されているため、それに合わせるのがよさそうです。

あと、以下のようにしたいです。
- Deploymentを -> Deploymentのロールアウトを
- 停止 -> 一時停止

> [Pause the rollout of a Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#pausing-and-resuming-a-deployment) to apply multiple fixes to its PodTemplateSpec and then resume it to start a new rollout.

* ロールアウトが停止したサインとして、[Deploymentのステータスを利用する](#deployment-status)。
* 今後必要としない[古いReplicaSetのクリーンアップを行う](#clean-up-policy)。

## Deploymentの作成 {#creating-a-deployment}

以下は、Deploymentの例です。

> これは`nginx`Podのレプリカを3つ持つReplicaSetを作成します:

Reviewer 2: ```suggestion
これは`nginx` Podのレプリカを3つ持つReplicaSetを作成します:
```
翻訳ガイドに沿って修正お願いします。
https://kubernetes.io/ja/docs/contribute/localization/#:~:text=%E3%83%90%E3%83%83%E3%82%AF%E3%82%AF%E3%82%A9%E3%83%BC%E3%83%88%60%E3%81%A7%E5%9B%B2%E3%81%BE%E3%82%8C%E3%81%9F%E8%8B%B1%E5%8D%98%E8%AA%9E%E3%81%A8%E3%80%81%E3%81%9D%E3%81%AE%E5%89%8D%E5%BE%8C%E3%81%AE%E8%8B%B1%E5%8D%98%E8%AA%9E%E3%81%AE%E9%96%93%E3%81%AB%E3%81%AF%E5%8D%8A%E8%A7%92%E3%82%B9%E3%83%9A%E3%83%BC%E3%82%B9%E3%81%8C%E5%BF%85%E8%A6%81

{{% code_sample file="controllers/nginx-deployment.yaml" %}}

この例では、

* `.metadata.name`フィールドで示されるとおり、`nginx-deployment`という名前のDeploymentが作成されます。
  この名前は、後で作成されるReplicaSetやPodの名前のもとになります。
  詳しくは[Deployment Specの記述](#writing-a-deployment-spec)を参照してください。
* このDeploymentは`.spec.replicas`フィールドで示されるとおり、3つのレプリカPodを作成するReplicaSetを作成します。
* `.spec.selector`フィールドは、作成されるReplicaSetが管理対象のPodをどのように見つけるかを定義します。
  ここでは、Podテンプレートで定義されたラベル(`app: nginx`)を選択しています。

>   しかし、Podテンプレート自体がルールを満たす限り、より洗練された選択ルールも可能です。

Reviewer: ```suggestion
  ただし、Podテンプレート自体がルールを満たす限り、より高度な選択ルールの指定も可能です。
```
次のようにすることで、元の翻訳文からもう少し改善できると思いました。
- 前の文を否定するほどでもない、断り書きの文なので、「ただし」に変更。
- sophisticatedは、一般的には「洗練された」と訳しますが、ここでは、単純なキーバリューの一致による条件指定だけでなく、集合ベースの条件指定もできるという技術的な話題であるため、「高度な」に変更。

> However, more sophisticated selection rules are possible, as long as the Pod template itself satisfies the rule.

Author: 以前も何かの翻訳で同じ指摘をいただいた気がします。修正します。

Reviewer: そういえば、richer behaiviorを、「より豊かな動作」から「より高度な動作」にする[コメント](https://github.com/kubernetes/website/pull/53598#discussion_r2798604663)をしたのを思い出しました。

  {{< note >}}

>   `.spec.selector.matchLabels`フィールドはキーバリューペアのマップです。

Reviewer: ```suggestion
  `.spec.selector.matchLabels`フィールドは{key,value}ペアのマップです。
```
> The .spec.selector.matchLabels field is a map of {key,value} pairs.

原文の{key,value}の表記に合わせておきましょうか。

>   `matchLabels`マップにおいて、{key, value}というペアは、keyというフィールドの値が"key"で、その演算子が"In"で、値の配列が"value"のみ含むような`matchExpressions`の要素と等しくなります。

Reviewer: ```suggestion
  `matchLabels`マップ内の単一の{key,value}は、`key`フィールドの値が"key"、`operator`が"In"、`values`配列には"value"のみを含む`matchExpressions`の要素と等しくなります。
```
より原文に近い翻訳にしました。

> A single {key,value} in the `matchLabels` map is equivalent to an element of `matchExpressions`, whose `key` field is "key", the `operator` is "In", and the `values` array contains only "value".

>   `matchLabels`と`matchExpressions`の両方の要件を、条件に一致させるために満たす必要があります。

Reviewer: ```suggestion
  条件に一致させるためには、`matchLabels`と`matchExpressions`の両方の要件を満たす必要があります。
```
些細なことですが、順番を入れ替えると読みやすくなりますね。

  {{< /note >}}

* `.spec.template`フィールドは、以下のサブフィールドを持ちます:
  * Podは`.metadata.labels`フィールドによって指定された`app: nginx`というラベルがつけられます。

>   * PodTemplate、または`.spec`フィールドは、Podが`nginx`という名前で[Docker Hub](https://hub.docker.com/)にある`nginx`のバージョン1.14.2が動くコンテナを1つ動かすことを示します。

Reviewer: ```suggestion
  * Podテンプレートの仕様、または`.spec`フィールドは、Podが`nginx`という名前で[Docker Hub](https://hub.docker.com/)にある`nginx`のバージョン1.14.2が動くコンテナを1つ動かすことを示します。
```
PodTemplate -> Podテンプレートの仕様

> The Pod template's specification, or `.spec` field ...

3. Deploymentのロールアウトステータスを確認するために、`kubectl rollout status deployment/nginx-deployment`を実行してください。

   コマンドの実行結果は以下のとおりです:

   ```
   Waiting for rollout to finish: 2 out of 3 new replicas have been updated...
   deployment "nginx-deployment" successfully rolled out
   ```

4. 数秒後、再度`kubectl get deployments`を実行してください。
   コマンドの実行結果は以下のとおりです:

   ```
   NAME               READY   UP-TO-DATE   AVAILABLE   AGE
   nginx-deployment   3/3     3            3           18s
   ```

   Deploymentが3つ全てのレプリカを作成して、全てのレプリカが最新(Podが最新のPodテンプレートを含んでいる)になり、利用可能となっていることを確認してください。

5. Deploymentによって作成されたReplicaSet(`rs`)を確認するには`kubectl get rs`を実行してください。
   コマンドの実行結果は以下のとおりです:

   ```
   NAME                          DESIRED   CURRENT   READY   AGE
   nginx-deployment-75675f5897   3         3         3       18s
   ```

   ReplicaSetの出力には次のフィールドが表示されます。

   * `NAME`は、名前空間内にあるReplicaSetの名前の一覧です。
   * `DESIRED`は、アプリケーションの理想的な _レプリカ_ の値です。
     これはDeploymentを作成したときに定義したもので、これが _理想的な状態_ と呼ばれるものです。
   * `CURRENT`は現在実行されているレプリカの数です。
   * `READY`は、ユーザーが使用できるアプリケーションのレプリカの数です。
   * `AGE`は、アプリケーションが稼働してからの時間です。

   ReplicaSetの名前は常に`[Deployment名]-[ハッシュ]`という形式になることに注意してください。
   この名前は、作成されるPodの名前のもとになります。

>    この`HASH`の文字列は、ReplicaSetの`pod-template-hash`ラベルと同じです。

Reviewer: ```suggestion
   この`ハッシュ`の文字列は、ReplicaSetの`pod-template-hash`ラベルと同じです。
```
`HASH`は、前の段落にある`[Deployment名]-[ハッシュ]`のことなので、表記を合わせたいです。

> 6. 各Podにラベルが自動的に付けられるのを確認するには`kubectl get pods --show-labels`を実行してください。

Reviewer: ```suggestion
6. 各Podに対して自動的に生成されたラベルを確認するには、`kubectl get pods --show-labels`を実行してください。
```
読みやすくしました。

> To see the labels automatically generated for each Pod, run `kubectl get pods --show-labels`.

   ```

   もしくは

   ```
   deployment "nginx-deployment" successfully rolled out
   ```

更新されたDeploymentのさらなる情報を取得するには、以下を確認してください:

* ロールアウトが成功したあと、`kubectl get deployments`を実行してDeploymentを確認できます。
  実行結果は以下のとおりです:

  ```
  NAME               READY   UP-TO-DATE   AVAILABLE   AGE
  nginx-deployment   3/3     3            3           36s
  ```

* Deploymentが新しいReplicaSetを作成してPodを更新させたり、新しいReplicaSetのレプリカを3にスケールアップさせたり、古いReplicaSetのレプリカを0にスケールダウンさせるのを確認するには`kubectl get rs`を実行してください。

  ```shell
  kubectl get rs
  ```

  実行結果は以下のとおりです:

  ```
  NAME                          DESIRED   CURRENT   READY   AGE
  nginx-deployment-1564180365   3         3         3       6s
  nginx-deployment-2035384211   0         0         0       36s
  ```

* `get pods`を実行すると、新しいPodのみ確認できます:

  ```shell
  kubectl get pods
  ```

  実行結果は以下のとおりです:

  ```
  NAME                                READY     STATUS    RESTARTS   AGE
  nginx-deployment-1564180365-khku8   1/1       Running   0          14s
  nginx-deployment-1564180365-nacti   1/1       Running   0          14s
  nginx-deployment-1564180365-z9gth   1/1       Running   0          14s
  ```

>   次にPodを更新させたいときは、DeploymentのPodテンプレートを再度更新するだけです。

Reviewer: ```suggestion
  次回これらのPodを更新させたいときは、DeploymentのPodテンプレートを再度更新するだけです。
```
手順を説明するときの、「次に○○をします」というような場合の「次に」にも読めてしまうのを避けた表現にしました。

文全体で見れば誤解しないとは思いますが、私は頭だけ読んだときに一瞬そう解釈してしまったので…。

> Next time you want to update these Pods, you only need to update the Deployment's Pod template again.

  Deploymentは、Podが更新されている間に特定の数のPodのみ停止状態になることを保証します。
  デフォルトでは、目標とするPod数の少なくとも75%が稼働状態であることを保証します(25% max unavailable)。

  また、DeploymentはPodが更新されている間に、目標とするPod数を特定の数まで超えてPodを稼働させることを保証します。

>   デフォルトでは、目標とするPod数に対して最大でも125%を超えてPodを稼働させることを保証します(25% max surge)。

Reviewer: ```suggestion
  デフォルトでは、目標とするPod数に対して最大でも125%までのPodを稼働させることを保証します(25% max surge)。
```
この説明は、「目標とするPod数(100%)を超えても125%までにする」という意味のはずなので、「125%を超えて」は「125%までの」に変更するのがよさそうです。

> By default, it ensures that at most 125% of the desired number of Pods are up (25% max surge).

  ロールアウトがうまくいかないときは、[Deploymentのステータスを読んでください](#deployment-status)。

* 古いレプリカ数(`nginx-deployment-1564180365`と`nginx-deployment-2035384211`のレプリカ数を合計した数)が3であり、新しいレプリカ数(`nginx-deployment-3066724191`のレプリカ数)が1であることを確認できます。

  ```shell
  kubectl get rs
  ```

  実行結果は以下のとおりです:

  ```
  NAME                          DESIRED   CURRENT   READY   AGE
  nginx-deployment-1564180365   3         3         3       25s
  nginx-deployment-2035384211   0         0         0       36s
  nginx-deployment-3066724191   1         1         0       6s
  ```

* 作成されたPodを確認していると、新しいReplicaSetによって作成された1つのPodはコンテナイメージ取得に失敗し続けているのがわかります。

  ```shell
  kubectl get pods
  ```

  実行結果は以下のとおりです:

  ```
  NAME                                READY     STATUS             RESTARTS   AGE
  nginx-deployment-1564180365-70iae   1/1       Running            0          25s
  nginx-deployment-1564180365-jbqqo   1/1       Running            0          25s
  nginx-deployment-1564180365-hysrc   1/1       Running            0          25s
  nginx-deployment-3066724191-08mng   0/1       ImagePullBackOff   0          6s
  ```

  {{< note >}}
  Deploymentコントローラーは、この悪い状態のロールアウトを自動的に停止し、新しいReplicaSetのスケールアップを止めます。

>   これはユーザーが指定したローリングアップデートに関するパラメーター(特に`maxUnavailable`)に依存します。

Reviewer: ```suggestion
  これはユーザーが指定したrollingUpdateのパラメーター(特に`maxUnavailable`)に依存します。
```
rollingUpdateは、以下のように記述する固有の名称のため、訳さずそのままがよいでしょうね。
```
rollingUpdate:
  maxUnavailable: 1
```
> This depends on the rollingUpdate parameters (maxUnavailable specifically) that you have specified.

     Normal  ScalingReplicaSet   11m   deployment-controller  Scaled down replica set nginx-deployment-75675f5897 to 2
     Normal  ScalingReplicaSet   11m   deployment-controller  Scaled up replica set nginx-deployment-c4747d96c to 2
     Normal  ScalingReplicaSet   11m   deployment-controller  Scaled down replica set nginx-deployment-75675f5897 to 1
     Normal  ScalingReplicaSet   11m   deployment-controller  Scaled up replica set nginx-deployment-c4747d96c to 3
     Normal  ScalingReplicaSet   11m   deployment-controller  Scaled down replica set nginx-deployment-75675f5897 to 0
     Normal  ScalingReplicaSet   11m   deployment-controller  Scaled up replica set nginx-deployment-595696685f to 1
     Normal  DeploymentRollback  15s   deployment-controller  Rolled back deployment "nginx-deployment" to revision 2
     Normal  ScalingReplicaSet   15s   deployment-controller  Scaled down replica set nginx-deployment-595696685f to 0
   ```

## Deploymentのスケーリング {#scaling-a-deployment}

以下のコマンドを実行させてDeploymentをスケールできます:

```shell
kubectl scale deployment/nginx-deployment --replicas=10
```

実行結果は以下のとおりです:

```
deployment.apps/nginx-deployment scaled
```

> クラスターで[水平Podオートスケーリング](/docs/tasks/run-application/horizontal-pod-autoscale/)が有効になっていると仮定すると、Deploymentのオートスケーラーを設定し、既存のPodのCPU使用率に基づいて、稼働させたいPodの最小数と最大数を選択できます。

Reviewer 2: リンクが英語版と違っているのでご確認お願いします。
https://github.com/kubernetes/website/blob/main/content/en/docs/concepts/workloads/controllers/deployment.md?plain=1#L647
/docs/concepts/workloads/autoscaling/horizontal-pod-autoscale/

```shell
kubectl autoscale deployment/nginx-deployment --min=10 --max=15 --cpu-percent=80%
```

実行結果は以下のとおりです:

```
deployment.apps/nginx-deployment scaled
```

### 比例スケーリング {#proportional-scaling}

Deploymentのローリングアップデートは、同時に複数のバージョンのアプリケーションを稼働させることをサポートします。

> ユーザーやオートスケーラーが、ロールアウトの途中(進行中もしくは一時停止中)にあるローリングアップデート方式のDeploymentをスケーリングすると、Deploymentコントローラーはリスクを軽減するために、既存のアクティブなReplicaSet(Podを持つReplicaSet)内で追加されるレプリカのバランスを取ります。

Reviewer: ```suggestion
RollingUpdate Deploymentは、同時に複数のバージョンのアプリケーションを稼働させることをサポートします。
ユーザーやオートスケーラーが、ロールアウトの途中(進行中もしくは一時停止中)にあるRollingUpdate Deploymentをスケーリングすると、Deploymentコントローラーはリスクを軽減するために、既存のアクティブなReplicaSet(Podを持つReplicaSet)内で追加されるレプリカのバランスを取ります。
```
RollingUpdateも、`type: RollingUpdate`のように指定する固有の名称のため、そのままにしましょうか。

> RollingUpdate Deployments support running multiple versions of an application at the same time. When you or an autoscaler scales a RollingUpdate Deployment that is in the middle of a rollout (either in progress or paused), the Deployment controller balances the additional replicas in the existing active ReplicaSets (ReplicaSets with Pods) in order to mitigate risk.

これを*比例スケーリング* と呼びます。

レプリカ数が10、[maxSurge](#max-surge)=3、[maxUnavailable](#max-unavailable)=2であるDeploymentが稼働している例です。

* Deployment内で10のレプリカが稼働していることを確認します:

  ```shell
  kubectl get deploy
  ```

  実行結果は以下のとおりです:

  ```
  NAME                 DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
  nginx-deployment     10        10        10           10          50s
  ```

> * クラスター内で、解決できない新しいイメージに更新します:

Reviewer 2: ```suggestion
* クラスター内からは、解決できない新しいイメージに更新します:
```
元々の訳ですが、`from inside the cluster`なので「クラスター内から」とした方が正確な訳になるかと思いました。

  ```shell
  kubectl set image deployment/nginx-deployment nginx=nginx:sometag
  ```

  実行結果は以下のとおりです:

  ```
  deployment.apps/nginx-deployment image updated
  ```

* イメージの更新は新しいReplicaSet nginx-deployment-1989198191へのロールアウトを開始させます。
  しかしロールアウトは、上述した`maxUnavailable`の要求によりブロックされます。
  ここでロールアウトのステータスを確認します:

  ```shell
  kubectl get rs
  ```

  実行結果は以下のとおりです:

  ```
  NAME                          DESIRED   CURRENT   READY     AGE
  nginx-deployment-1989198191   5         5         0         9s
  nginx-deployment-618515232    8         8         8         1m
  ```

* 次にDeploymentをスケーリングするための新しい要求が発生します。
  オートスケーラーはDeploymentのレプリカ数を15に増やします。
  Deploymentコントローラーは新しい5つのレプリカをどこに追加するか決める必要がでてきます。
  比例スケーリングを使用していない場合、5つのレプリカは全て新しいReplicaSetに追加されます。
  比例スケーリングでは、追加されるレプリカは全てのReplicaSetに分散されます。

>   比例割合が大きいものはレプリカ数の大きいReplicaSetとなり、比例割合が低いときはレプリカ数の小さいReplicaSetとなります。

Reviewer: ```suggestion
  比例割合が大きいものはレプリカ数の大きいReplicaSetとなり、比例割合が小さいときはレプリカ数の小さいReplicaSetとなります。
```
比例割合が「大きい」に対して、「小さい」として表現を合わせています。

* 不十分なクォータ
* ReadinessProbeの失敗
* コンテナイメージの取得エラー
* 不十分なパーミッション
* リソースリミットのレンジ
* アプリケーションランタイムの設定の不備

このような状況を検知する1つの方法として、Deploymentのリソース定義でデッドラインのパラメーターを指定します([`.spec.progressDeadlineSeconds`](#progress-deadline-seconds))。
`.spec.progressDeadlineSeconds`は、Deploymentの更新が停止したことを(Deploymentのステータスで)示す前にDeploymentコントローラーが待つ秒数を示します。

以下の`kubectl`コマンドでリソース定義に`progressDeadlineSeconds`を設定します。
これはDeploymentの更新が止まってから10分後に、コントローラーが進行の停止を通知させるためです。

```shell
kubectl patch deployment/nginx-deployment -p '{"spec":{"progressDeadlineSeconds":600}}'
```

実行結果は以下のとおりです:

```
deployment.apps/nginx-deployment patched
```

デッドラインを超過すると、Deploymentコントローラーは以下の属性を持つDeploymentConditionをDeploymentの`.status.conditions`に追加します:

* `type: Progressing`
* `status: "False"`
* `reason: ProgressDeadlineExceeded`

このConditionは、`ReplicaSetCreateError`のような理由により早期に失敗し、ステータス値`"False"`に設定されることもあります。
また、Deploymentのロールアウトが完了すると、デッドラインは考慮されなくなります。

ステータスの状態に関するさらなる情報は[Kubernetes APIの規則](https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties)を参照してください。

{{< note >}}
Kubernetesは、`reason: ProgressDeadlineExceeded`のステータス状態を報告する以外に、停止状態のDeploymentに対して何のアクションも実行しません。

> 高レベルのオーケストレーターはこれを利用して、状態に応じて行動できます。

Reviewer: ```suggestion
上位レベルのオーケストレーターはこれを利用して、状態に応じて行動できます。
```
higher levelを、上位層の意味として誤解なく読めるようにしました。

例えば、前のバージョンへのDeploymentのロールバックが挙げられます。
{{< /note >}}

{{< note >}}

> Deploymentを一時停止すると、Kubernetesは指定したデッドラインを超えたかどうかチェックしません。

Reviewer: ```suggestion
Deploymentのロールアウトを一時停止しても、Kubernetesは指定したデッドラインを超えたかどうかチェックしません。
```
「ロールアウト」を足しました。
また、「一時停止しても」とすると、日本語として読みやすく感じました。

> If you pause a Deployment rollout, ...

Deploymentを使って一部のユーザーやサーバーに対してリリースのロールアウトをしたい場合は、[リソースの管理](/docs/concepts/workloads/management/#canary-deployments)に記載されているカナリアパターンに従って、リリース毎に1つずつ、複数のDeploymentを作成できます。

## Deployment Specの記述 {#writing-a-deployment-spec}

他の全てのKubernetesの設定と同様に、Deploymentは`.apiVersion`、`.kind`、`.metadata`フィールドを必要とします。
設定ファイルの利用に関する一般的な情報については、[アプリケーションのデプロイ](/docs/tasks/run-application/run-stateless-application-deployment/)、コンテナの設定、および[リソースを管理するためのkubectlの使用](/docs/concepts/overview/working-with-objects/object-management/)のドキュメントを参照してください。

コントロールプレーンがDeployment用に新しいPodを作成するとき、Deploymentの`.metadata.name`がそれらのPodの名前付けのもとの一部になります。
Deploymentの名前は有効な[DNSサブドメイン名](/docs/concepts/overview/working-with-objects/names#dns-subdomain-names)である必要がありますが、これはPodのホスト名に予期しない結果をもたらすことがあります。
最良の互換性のために、名前はより制限の厳しい[DNSラベル名](/docs/concepts/overview/working-with-objects/names#dns-label-names)のルールに従う必要があります。

Deploymentは[`.spec`セクション](https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status)も必要とします。

### Podテンプレート {#pod-template}

`.spec.template`と`.spec.selector`は`.spec`における唯一の必須フィールドです。

`.spec.template`は[Podテンプレート](/docs/concepts/workloads/pods/#pod-templates)です。

> これは`.spec`内でネストされていることと、`apiVersion`や`kind`を持たないことを除いては{{< glossary_tooltip text="Pod" term_id="pod" >}}と全く同じスキーマとなります。

Reviewer 2: ```suggestion
これは`.spec`内でネストされていることと、`apiVersion`や`kind`を持たないことを除いては{{< glossary_tooltip text="Pod" term_id="pod" >}}と全く同じスキーマです。
```
元の訳で恐縮ですが、翻訳ガイドに沿って「です」にするのはいかがでしょうか。
https://kubernetes.io/ja/docs/contribute/localization/#:~:text=%E7%89%B9%E3%81%AB%E3%80%81%E3%80%8C%E3%80%9C%E3%81%AB%E3%81%AA%E3%82%8A%E3%81%BE%E3%81%99%E3%80%8D%E3%80%8C%E3%80%9C%E3%81%A8%E3%81%AA%E3%82%8A%E3%81%BE%E3%81%99%E3%80%8D%E3%81%A8%E3%81%84%E3%81%86%E8%A1%A8%E7%8F%BE%E3%81%AF%E3%80%8C%E3%80%9C%E3%81%A7%E3%81%99%E3%80%8D%E3%81%AE%E6%96%B9%E3%81%8C%E9%81%A9%E5%88%87%E3%81%AA%E5%A0%B4%E5%90%88%E3%81%8C%E5%A4%9A%E3%81%84%E3%81%9F%E3%82%81%E6%B3%A8%E6%84%8F

Podの必須フィールドに加えて、Deployment内のPodテンプレートでは適切なラベルと適切な再起動ポリシーを設定しなくてはなりません。
ラベルは他のコントローラーと重複しないようにしてください。
ラベルについては、[セレクター](#selector)を参照してください。

[`.spec.template.spec.restartPolicy`](/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy)が`Always`に等しいときのみ許可されます。
これは指定されていない場合のデフォルト値です。

### レプリカ数 {#replicas}

`.spec.replicas`は理想的なPodの数を指定するオプションのフィールドです。
デフォルトは1です。

`kubectl scale deployment deployment --replicas=X`のようにDeploymentを手動でスケールし、その後マニフェストに基づいてそのDeploymentを更新した場合(例えば`kubectl apply -f deployment.yaml`を実行した場合)、そのマニフェストの適用によって、以前に行った手動のスケーリングは上書きされます。

> [HorizontalPodAutoscaler](/docs/tasks/run-application/horizontal-pod-autoscale/)(または水平スケーリングのための同様のAPI)がDeploymentのスケーリングを管理している場合、`.spec.replicas`を設定しないでください。

Reviewer 2: 気付けておらずすみません。
リンクこちらも修正お願いします。

`apps/v1`バージョンにおいて、`.spec.selector`と`.metadata.labels`が指定されていない場合、`.spec.template.metadata.labels`の値に初期化されません。
そのため`.spec.selector`と`.metadata.labels`を明示的に指定する必要があります。
また`apps/v1`のDeploymentにおいて`.spec.selector`は作成後に不変になります。

Deploymentのテンプレートが`.spec.template`と異なるラベルを持つPodや、`.spec.replicas`の値を超えて稼働しているPodがある場合、Deploymentはセレクターに一致するラベルを持つPodを削除することがあります。
Podの数が理想状態より少ない場合、Deploymentは`.spec.template`をもとに新しいPodを作成します。

{{< note >}}
このセレクターに一致するラベルを持つPodを、直接作成したり、他のDeploymentを作成したり、ReplicaSetやReplicationControllerのような他のコントローラーを作成したりして作るべきではありません。
作成してしまうと、最初のDeploymentがこれらの他のPodを作成したとみなしてしまいます。
ただし、Kubernetesはこれを行うことを止めません。
{{< /note >}}

セレクターが重複する複数のコントローラーを持つとき、そのコントローラーは互いに競合状態となり、正しくふるまいません。

### 更新戦略 {#strategy}

`.spec.strategy`は古いPodから新しいPodに置き換える際の更新戦略を指定します。
`.spec.strategy.type`は"Recreate"もしくは"RollingUpdate"を指定できます。
デフォルトは"RollingUpdate"です。

#### Deploymentの再作成 {#recreate-deployment}

`.spec.strategy.type==Recreate`と指定されているとき、既存の全てのPodは新しいPodが作成される前に削除されます。

{{< note >}}
これは、アップグレードの際に、Podの作成より前にPodが停止されることを保証するだけです。
Deploymentを更新する場合、古いリビジョンのPodは全てすぐに停止されます。
削除に成功するまでは、新しいリビジョンのPodは作成されません。

> 手動でPodを削除すると、ライフサイクルがReplicaSetに制御されているのですぐに置き換えが実施されます(たとえ古いPodがまだ停止中のステータスでも)。

Reviewer: ```suggestion
手動でPodを削除すると、ライフサイクルはReplicaSetによって制御されているため、すぐに置き換えが実施されます(たとえ古いPodがまだ停止中のステータスでも)。
```
- ライフサイクルが -> ライフサイクルは
- ので -> ため、

> Podに"高々この程度の"保証を求めるならば[StatefulSet](/docs/concepts/workloads/controllers/statefulset/)の使用を検討してください。

Reviewer: ```suggestion
Podに対して「最大稼働数」の保証が必要であれば、[StatefulSet](/docs/concepts/workloads/controllers/statefulset/)の使用を検討してください。
```
原文では、「設定されたレプリカ数を、一時的であっても絶対に超えない保証」が必要なケースのことを言っているように思います。

at mostは、イディオムとしての「高々」というよりは、at most oneで「最大で1つ」のような、最大数のことを指しているのでしょうね。

> If you need an "at most" guarantee for your Pods, you should consider using a [StatefulSet](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/).

{{< /note >}}

#### Deploymentのローリングアップデート {#rolling-update-deployment}

`.spec.strategy.type==RollingUpdate`と指定されているとき、DeploymentはローリングアップデートによりPodを更新します(古いReplicaSetを段階的にスケールダウンし、新しいReplicaSetをスケールアップします)。
ローリングアップデートの処理をコントロールするために`maxUnavailable`と`maxSurge`を指定できます。

##### Max Unavailable {#max-unavailable}

`.spec.strategy.rollingUpdate.maxUnavailable`はオプションのフィールドで、更新処理において利用不可となる最大のPod数を指定します。
値は絶対値(例: 5)を指定するか、理想状態のPodのパーセンテージを指定します(例: 10%)。

> パーセンテージを指定した場合、絶対値は小数切り捨てされて計算されます。

Reviewer: ```suggestion
値は絶対数(例: 5)を指定するか、理想状態のPodのパーセンテージを指定します(例: 10%)。
パーセンテージを指定した場合、絶対数は小数切り捨てされて計算されます。
```
「絶対値」は負の数がない表現とも受け取られるため、absolute numberは「絶対数」とするのがよいのではと思いました。
ここでは、比率ではなく指定した数値をそのまま個数として扱うことを言いたいのでしょう。

`.spec.strategy.rollingUpdate.maxSurge`が0に指定されている場合、この値を0にできません。
デフォルトでは25%です。

例えば、この値が30%と指定されているとき、ローリングアップデートが開始すると古いReplicaSetはすぐに理想状態の70%にスケールダウンされます。
一度新しいPodが稼働できる状態になると、古いReplicaSetはさらにスケールダウンされ、続いて新しいReplicaSetがスケールアップされます。
この間、利用可能なPodの総数は理想状態のPodの少なくとも70%以上になるように保証されます。

##### Max Surge {#max-surge}

`.spec.strategy.rollingUpdate.maxSurge`はオプションのフィールドで、理想状態のPod数を超えて作成できる最大のPod数を指定します。
値は絶対値(例: 5)を指定するか、理想状態のPodのパーセンテージを指定します(例: 10%)。

> パーセンテージを指定した場合、絶対値は小数切り上げで計算されます。

Reviewer: ```suggestion
値は絶対数(例: 5)を指定するか、理想状態のPodのパーセンテージを指定します(例: 10%)。
パーセンテージを指定した場合、絶対数は小数切り上げで計算されます。
```
同上。

        - containerPort: 80
  strategy:
    type: RollingUpdate
    rollingUpdate:
      maxSurge: 1
 ```

{{% /tab %}}
{{% tab name="Hybrid" %}}

 ```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  labels:
    app: nginx
spec:
  replicas: 3
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:1.14.2
        ports:
        - containerPort: 80
  strategy:
    type: RollingUpdate
    rollingUpdate:
      maxSurge: 1
      maxUnavailable: 1
 ```

{{% /tab %}}
{{< /tabs >}}

### Progress Deadline Seconds {#progress-deadline-seconds}

`.spec.progressDeadlineSeconds`はオプションのフィールドで、システムがDeploymentの[進行に失敗](#failed-deployment)したと報告するまでに、Deploymentが進行するのを待つ秒数を指定します。
失敗は、リソースのステータスにおいて`type: Progressing`、`status: "False"`、`reason: ProgressDeadlineExceeded`を持つConditionとして表面化します。
DeploymentコントローラーはDeploymentのリトライを続けます。
デフォルト値は600です。

このフィールドを指定する場合、`.spec.minReadySeconds`より大きい値にする必要があります。

### Min Ready Seconds {#min-ready-seconds}

`.spec.minReadySeconds`はオプションのフィールドで、新しく作成されたPodが利用可能となるために、最低どれくらいの秒数コンテナがクラッシュすることなく稼働し続ければよいかを指定するものです。

> デフォルトでは0です(Podは作成されるとすぐに利用可能と判断されます)。

Reviewer: ```suggestion
`.spec.minReadySeconds`はオプションのフィールドで、新しく作成されたPodが利用可能となるために、いずれのコンテナもクラッシュすることなく準備完了の状態を維持すべき最小秒数を指定するものです。
デフォルトでは0です(Podは準備完了になるとすぐに利用可能と判断されます)。
```
any of its containersを「いずれのコンテナも」としました。
readyの訳し方も、説明の正確性の面で気になったため変えました。

> `.spec.minReadySeconds` is an optional field that specifies the minimum number of seconds for which a newly created Pod should be ready without any of its containers crashing, for it to be considered available. This defaults to 0 (the Pod will be considered available as soon as it is ready).

> Podが利用可能と判断される場合についてさらに学ぶには、[Container Probes](/docs/concepts/workloads/pods/pod-lifecycle/#container-probes)を参照してください。

Reviewer: ```suggestion
Podが利用可能と判断される場合についてさらに学ぶには、[コンテナのProbe](/docs/concepts/workloads/pods/pod-lifecycle/#container-probes)を参照してください。
```
日本語ページの見出しに合わせました。

### 終了中のPod {#terminating-pods}

{{< feature-state feature_gate_name="DeploymentReplicaSetTerminatingReplicas" >}}

終了中(terminating)のPodは、[APIサーバー](/docs/reference/command-line-tools-reference/kube-apiserver/)と[kube-controller-manager](/docs/reference/command-line-tools-reference/kube-controller-manager/)で`DeploymentReplicaSetTerminatingReplicas`[フィーチャーゲート](/docs/reference/command-line-tools-reference/feature-gates/)が有効になっている場合にのみ確認できます。

削除やスケールダウンによって終了中になったPodは、終了するまでに長い時間がかかることがあり、その間に追加のリソースを消費する可能性があります。
その結果、全Podの合計数が一時的に`.spec.replicas`を超えることがあります。
終了中のPodは、Deploymentの`.status.terminatingReplicas`フィールドを使用して追跡できます。

### リビジョン履歴の保持上限 {#revision-history-limit}

Deploymentのリビジョン履歴は、Deploymentが管理するReplicaSetに保持されています。

`.spec.revisionHistoryLimit`はオプションのフィールドで、ロールバックを可能にするために保持する古いReplicaSetの数を指定します。
この古いReplicaSetは`etcd`内のリソースを消費し、`kubectl get rs`の出力結果を見にくくします。
Deploymentの各リビジョンの設定はReplicaSetに保持されます。
このため一度古いReplicaSetが削除されると、そのリビジョンのDeploymentにロールバックすることができなくなります。
デフォルトでは10もの古いReplicaSetが保持されますが、この値の最適値は新しいDeploymentの更新頻度と安定性に依存します。

> さらに詳しく言うと、この値を0にすると、0のレプリカを持つ古い全てのReplicaSetが削除されます。

Reviewer: ```suggestion
さらに詳しく言うと、この値を0にすると、レプリカ数が0の古いReplicaSetが全て削除されます。
```
0の場合は「持つ」とはあまり言わないため、表現を変えました。
